tools/generate_code_cache.js

```diff
@@ -0,0 +1,97 @@
+'use strict';
+
+const natives = process.binding('natives');
+const vm = require('vm');
+const fs = require('fs');
+let count = 0;
+
+function human(num) {
+  if (num < 1024) {
+    return `${(num).toFixed(2)}B`;
+  } else if (num < 1024 ** 2) {
+    return `${(num / 1024).toFixed(2)}KB`;
+  } else if (num < 1024 ** 3) {
+    return `${(num / (1024 ** 2)).toFixed(2)}MB`;
+  } else {
+    return `${(num / (1024 ** 3)).toFixed(2)}GB`;
+  }
+}
+
+/**
+ * @param {string} key
+ * @param {Buffer} cache
+ */
+function getInitalizer(key, cache) {
+  const defName = key.replace(/\//g, '_').replace(/-/g, '_');
+  const def = `static uint8_t ${defName}_raw[] = {\n` +
+                     `${cache.join(',')}\n};`;
+  const initializer = `
+  v8::Local<v8::ArrayBuffer> ${defName}_ab =
+    v8::ArrayBuffer::New(isolate, ${defName}_raw, ${cache.length});
+  v8::Local<v8::Uint8Array> ${defName}_array =
+    v8::Uint8Array::New(${defName}_ab, 0, ${cache.length});
+  target->Set(context,
+              OneByteString(isolate, "${key}"),
+              ${defName}_array).FromJust();
+  `;
+  return {
+    def, initializer
+  };
+}
+
+const cacheDefs = [];
+const initializers = [];
+
+for (const key of Object.keys(natives)) {
+  if (key === 'config') continue;
+  const wrapper = [
+    '(function (exports, require, module, process) {',
+    '\n});'
+  ];
+  const code = wrapper[0] + natives[key] + wrapper[1];
+
+  const script = new vm.Script(code, {
+    filename: `${key}.js`,
+    produceCachedData: true
```

Comment:
This seems to use the old API to produce cache right after compile. This either compiles everything if specified to compiled eagerly, or only the toplevel function.

With the new API for which @devsnek is writing the binding, you could precisely include the functions you require for startup.

Maybe something to think about for the future :)

Reply:
@hashseed Yeah this is also why I think it's a good idea to return error information in that PR, for example if I run this script with `--inspect-brk` and open the devtools I would not be able to create any cache at all. I know it's because it's in debug mode but that's only because I've looked into the API before. If we want to open up tooling opportunities to users it would be nice to be able to tell then why the code cache cannot be created.

Reply:
> This either compiles everything if specified to compiled eagerly, or only the toplevel function.

FYI: We do not actually support eager compilation through the `vm.Script` API, so this is only the top-level function.

Reply:
@TimothyGu We should probably add support for the full range of `CompileOptions` to `ContextifyScript` (in another PR of course)